### PR TITLE
add proxy setting

### PR DESCRIPTION
**Pull request by LuckyTain**
Some endpoint block some IP address, add a aiohttp proxy setting to client. 

https://docs.aiohttp.org/en/stable/client_advanced.html

Hoyolab(oversea) block Chinese IP, a http proxy can help😀

### CODE REVIEW

**Comment by ashleney** [comment]
This can be done by overwriting the `create_session` method of your cookie manager. I like the idea of making a proxy easier though.

**Comment by LuckyTain** [comment]
> This can be done by overwriting the `create_session` method of your cookie manager. I like the idea of making a proxy easier though.

I have no idea how to overwriting the `create_session` method to make this work. 
In aiohttp doc, I only found I can set proxy inside session.request(proxy=)

File: genshin/client/components/base.py [hunk 1]
```diff
@@ -36,6 +37,7 @@ class BaseClient(abc.ABC):
     _lang: str
     _region: types.Region
     _default_game: typing.Optional[types.Game]
+    _proxy: typing.Optional[typing.Union[pydantic.AnyUrl, str]]
```
Comment (ashleney) [inline genshin/client/components/base.py:40]: This should ideally be an attribute of the cookie manager

File: genshin/client/components/base.py [hunk 2]
```diff
@@ -57,6 +60,12 @@ def __init__(
         self.authkey = authkey
         self.lang = lang
         self.region = region
+        if not isinstance(proxy, pydantic.AnyUrl):
+            try:
+                pydantic.parse_obj_as(pydantic.AnyUrl, proxy)
```
Comment (ashleney) [inline genshin/client/components/base.py:65]: `yarl.URL` is used instead of `pydantic`'s urls

File: genshin/client/components/base.py [hunk 3]
```diff
@@ -107,6 +117,15 @@ def region(self, region: str) -> None:
...
             self.lang = "zh-cn"
 
+    @property
+    def proxy(self) -> typing.Optional[typing.Union[pydantic.AnyUrl, str]]:
+        """The default proxy"""
+        return self._proxy
+
+    @proxy.setter
+    def proxy(self, proxy: typing.Optional[typing.Union[pydantic.AnyUrl, str]]) -> None:
+        self._proxy = proxy if proxy else None
```
Comment (ashleney) [inline genshin/client/components/base.py:127]: This does practically nothing in terms of validation, probably didn't even need to be a property at that point.

File: genshin/client/components/base.py [hunk 4]
```diff
@@ -263,6 +282,9 @@ async def request(
             if value is not None:
                 return value
 
+        """Set proxy"""
```
Comment (ashleney) [inline genshin/client/components/base.py:285]: This should be a comment, not a docstring.

**Review by ashleney — CHANGES_REQUESTED** [review]
Great idea but I think it could be implemented better. I'm probably gonna try to do this myself anyway.

**Comment by LuckyTain** [comment]
Thank you, I review the code and make some changes, I will open another pull requests

**Comment by ashleney** [comment]
Why another pull request? You can just edit the code here, can't you?

**Comment by GauravM512** [comment]
Yes just change your source , it'll change

**Comment by LuckyTain** [comment]
> Why another pull request? You can just edit the code here, can't you?

Because I create a new branch to rewrite, and I want to keep the old branch.
sorry for my lack knowledge of git

**Pull request closed without merging**